### PR TITLE
makes dependency on zlib optional

### PR DESCRIPTION
**Pull request by jangorecki**
Follow-up of #3872
It is NOT meant to solve Solaris issue, but general issue when zlib is not available.

---

before PR
```
docker run -it rocker/r-ver:3.5.3 Rscript -e 'install.packages("data.table", repos="https://Rdatatable.gitlab.io/data.table"); data.table::fwrite(iris, "iris.gz")'
```
```
*** pkg-config is not installed.
*** Compilation will now be attempted and if it works you can ignore this message. In
*** particular, this should be the case on Mac where zlib is built in.
*** However, if compilation fails, try 'locate zlib.h zconf.h' and ensure the zlib
*** development library is installed :
***   deb: zlib1g-dev (Debian, Ubuntu, ...)
***   rpm: zlib-devel (Fedora, EPEL, ...)
***   There is a zlib in brew for OSX but the built in zlib should work.
*** Note that zlib is required to compile R itself so you may find the advice in the R-admin
*** guide helpful regarding zlib. On Debian/Ubuntu, zlib1g-dev is a dependency of r-base as
*** shown by 'apt-cache showsrc r-base | grep ^Build-Depends | grep zlib', and therefore
*** 'sudo apt-get build-dep r-base' should be sufficient too.
*** To silence this message, please ensure that :
***   1) 'pkg-config --exists zlib' succeeds (i.e. $? -eq 0)
***   2) 'pkg-config --libs zlib' contains -lz
*** Compilation will now be attempted ...
...
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fopenmp -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=f
ormat-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fwrite.c -o fwrite.o
fwrite.c:10:49: fatal error: zlib.h: No such file or directory
 #include <zlib.h>      // for compression to .gz
```

---

after PR
```
docker run -it rocker/r-ver:3.5.3 Rscript -e 'install.packages("remotes", repos="https://cloud.r-project.org"); remotes::install_github("Rdatatable/data.table@nozlib"); data.table::fwrite(iris, "iris.gz")'
```
```
*** pkg-config is not installed.
*** Compilation will now be attempted and if it works you can ignore this message. In
*** particular, this should be the case on Mac where zlib is built in or pkg-config
*** is not installed. However, if compilation fails, try 'locate zlib.h zconf.h' and
*** ensure the zlib development library is installed :
***   deb: zlib1g-dev (Debian, Ubuntu, ...)
***   rpm: zlib-devel (Fedora, EPEL, ...)
***   There is a zlib in brew for OSX but the built in zlib should work.
*** Note that zlib is required to compile R itself so you may find the advice in the R-admin
*** guide helpful regarding zlib. On Debian/Ubuntu, zlib1g-dev is a dependency of r-base as
*** shown by 'apt-cache showsrc r-base | grep ^Build-Depends | grep zlib', and therefore
*** 'sudo apt-get build-dep r-base' should be sufficient too.
*** To silence this message, please ensure that :
***   1) 'pkg-config --exists zlib' succeeds (i.e. $? -eq 0)
***   2) 'pkg-config --libs zlib' contains -lz
*** Compilation will now be attempted ...
R CMD SHLIB supports OpenMP without any extra hint
*** Compilation without compression support in fwrite
...
Error in data.table::fwrite(iris, "iris.gz") : 
  Compression in fwrite uses zlib library which was not installed at the time when data.table was compiled. To enable compression support in fwrite one must install zlib and re-install data.table.
```

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -46,9 +48,9 @@ fi
 
 if [ $msg -ne 0 ]; then
   echo "*** Compilation will now be attempted and if it works you can ignore this message. In"
-  echo "*** particular, this should be the case on Mac where zlib is built in."
-  echo "*** However, if compilation fails, try 'locate zlib.h zconf.h' and ensure the zlib"
-  echo "*** development library is installed :"
+  echo "*** particular, this should be the case on Mac where zlib is built in or pkg-config"
+  echo "*** is not installed. However, if compilation fails, try 'locate zlib.h zconf.h' and"
+  echo "*** ensure the zlib development library is installed :"
   echo "***   deb: zlib1g-dev (Debian, Ubuntu, ...)"
```
Comment (MichaelChirico) [inline configure:54]: doesn't this turn off zlib support on mac by default

Comment (jangorecki) [reply configure:54]: If `pkg-config` is not available on Mac then yes, it turns off. Needs testing anyway.
There is a related comment above:
>  if pkg-config is not available then zlib will be disabled for higher chance of compilation success

Comment (jangorecki) [reply configure:54]: Note that branch was hit before when pkg-config was not installed, just this console message was not mentioning that.

Comment (jangorecki) [reply configure:54]: We could eventually try to check if zlib is available by attempting to compile dummy code, like we do for openmp. Docker image in your report didn't have either pkg-config and zlib. So then we could still compile with zlib support. For now I don't think it is worth. Messages are clear and points users in the proper direction. Installing pkg-config and/or zlib.

**Comment by jangorecki** [comment]
I put it on label High so we can have it merged at the beginning on a development cycle to be well tested before it will land on CRAN.

**Comment by mattdowle** [comment]
Changes I made in 08a2584:
* zlib runtime is still a requirement for those installing the binary package from CRAN. So reinstated the SystemRequirements line in DESCRIPTION. Not a problem because the zlib runtime is already required by R.  Further, `test.data.table()` will fail, and therefore so will `R CMD check`, if data.table has not been compiled with zlib headers because tests of `fwrite` compression will fail.
* `#define z_stream" empty line removed. I'm not sure that would work. I checked all uses of `z_stream` are inside `#ifndef NOZLIB blocks` blocks.
* distinction between library and headers made in the messages
* the error message now tells the user that reinstalling will output further guidance

**Pull request merged**